pkg/core/blockchain.go

```diff
@@ -524,34 +499,59 @@ func (bc *Blockchain) GetBlock(hash util.Uint256) (*Block, error) {
 	return block, nil
 }
 
+// getBlockFromStore returns Block by the given hash if it exists in the store.
+func getBlockFromStore(s storage.Store, hash util.Uint256) (*Block, error) {
+	key := storage.AppendPrefix(storage.DataBlock, hash.BytesReverse())
+	b, err := s.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	block, err := NewBlockFromTrimmedBytes(b)
+	if err != nil {
+		return nil, err
+	}
+	return block, err
+}
+
 // GetHeader returns data block header identified with the given hash value.
 func (bc *Blockchain) GetHeader(hash util.Uint256) (*Header, error) {
-	block, ok := bc.blockCache.GetBlock(hash)
-	if !ok {
-		b, err := bc.Get(storage.AppendPrefix(storage.DataBlock, hash.BytesReverse()))
-		if err != nil {
-			return nil, err
-		}
-		block, err = NewBlockFromTrimmedBytes(b)
+	header, err := getHeaderFromStore(bc.memStore, hash)
+	if err != nil {
+		header, err = getHeaderFromStore(bc.Store, hash)
 		if err != nil {
 			return nil, err
 		}
 	}
+	return header, err
+}
+
+// getHeaderFromStore returns Header by the given hash from the store.
+func getHeaderFromStore(s storage.Store, hash util.Uint256) (*Header, error) {
+	b, err := s.Get(storage.AppendPrefix(storage.DataBlock, hash.BytesReverse()))
+	if err != nil {
+		return nil, err
+	}
+	block, err := NewBlockFromTrimmedBytes(b)
+	if err != nil {
+		return nil, err
+	}
 	return block.Header(), nil
 }
 
 // HasTransaction return true if the blockchain contains he given
 // transaction hash.
 func (bc *Blockchain) HasTransaction(hash util.Uint256) bool {
-	if _, _, ok := bc.blockCache.GetTransaction(hash); ok {
-		return true
-	}
 	if bc.memPool.ContainsKey(hash) {
 		return true
 	}
+	return checkTransactionInStore(bc.memStore, hash) || checkTransactionInStore(bc.Store, hash)
```

Comment:
could be:
return
bc.memPool.ContainsKey(hash) ||
checkTransactionInStore(bc.memStore, hash) ||
checkTransactionInStore(bc.Store, hash) ||

Reply:
Not exactly the style I like, but as we already have `check() || check()`, probably it makes sense to go further with that.